### PR TITLE
Editor: Always render the 'Switch to Draft' button to avoid focus loss

### LINKED ISSUE #51901: Focus loss when switching to draft
### Description

When switching a Published or Scheduled post to Draft:
- The Switch to Draft button gets a `disabled` attribute while saving.
- This triggers an immediate focus loss, as the element that had focus is now not focusable. See https://github.com/WordPress/gutenberg/issues/15524
- A confirm dialog opens, focus is moved to the dialog.
- WHen pressing OK to save as a Draft, the dialog opens and the Switch to Draft is removed from the DOM.
- As such, the dialog can't restore focus on the control that opened it.
- A focus loss occurs.
- At this point, when pressing the Tab key, the tab sequence starts from scratch from the document root.

Animated GIF to illustrate:

switch to draft focus loss

Worth noting the focus loss occurs also in WordPress 6.2, However, the Switch to Draft button was placed in the editor top bar. While the focus loss still occurs, I'd say it wa less serious because focus landed not that far from where it was.

Instead, with the new placement of the Switch to Draft button in the settings panel, users are now forced to navigate almost the entire editor UI to go back where they were.

The expected behavior is that focus is returned to the UI control that opened the modal dialog. As such, the 'Switch to Draft' button should not be removed from the DOM. Ideally, it shoul dstay in the DOM, be still focusable, use an `aria-disabled` attribute and 'noop'.  See: ARIA Authoring Practices: [Focusability of disabled controls](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#focusabilityofdisabledcontrols).

### Step-by-step reproduction instructions

- Edit a Published or Scheduled post.
- Use the keyboard and navigate with the Tab key to the 'Switch to Draft' button in the Settings panel.
- Press Enter: a confirm dialog opens.
- Tab to the OK button and press Enter.
- The dialog closes.
- Observe the 'Switch to Draft' button is removed from the DOM.
- Press the Tab key.
- Observe the tab sequence starts from scratch from the Site icon button at the beginning of the document.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
Fixes #51901.

PR updates the `PostSwitchToDraftButton` component always to render to button, but using a disabled state when switching to draft isn't possible.

## Why?
See #51901.

> Instead, with the new placement of the Switch to Draft button in the settings panel, users are now forced to navigate almost the entire editor UI to go back where they were.

> The expected behavior is that focus is returned to the UI control that opened the modal dialog. As such, the 'Switch to Draft' button should not be removed from the DOM. Ideally, it shoul dstay in the DOM, be still focusable, use an aria-disabled attribute and 'noop'. See: ARIA Authoring Practices: [Focusability of disabled controls](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#focusabilityofdisabledcontrols).

## How?
* Updates button to use `aria-disabled` instead of `disabled` prop.
* Noops the `onClick` callback when the button is disabled.

## Testing Instructions
1. Open a published or scheduled post.
2. Navigate to the "Switch to Draft" button in document settings.
3. Switch the post to draft.
4. Verify that the button is visible, but has a disabled state.

### Testing Instructions for Keyboard
Same

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/240569/e830a441-7d75-4ea6-bd3d-5e37e4d99d05

### CODE REVIEW

**Comment by jasmussen** [comment]
Tricky one, obviously needs to be solved so thank you for the PR. It surfaces a shortcoming in this design, definitely — because it isn't ideal that "switch to draft" as a big honking always present button is visible on draft posts when it's entirely non functional. 

I suspect we need to revisit this entire panel and keeping this focus loss in mind so we can find a new design that covers all those bases. 

In the mean time, this does seem like the simplest path forward. The bigger issue here is the lack of a border on the [disabled state of the secondary button style](https://wordpress.github.io/gutenberg/?path=/story/components-button--secondary&args=disabled:!true). Can we/should we update that to include the border, can even be lighter if need be. This is mainly to maintain the box grid.

**Comment by Mamaduka** [comment]
100%  - It's not the prettiest solution, but it's a good intermediate one while working out the new design.

The editor has other actions with the same issues, e.g., the "block toolbar" unlock button. It would be great to have a general design guide on handling this tricky a11y and design challenges.

I also noticed a lack of focus styles. Do we have CSS/styles that I can reuse for this?

**Comment by jasmussen** [comment]
The block toolbar is a difficult challenge I'd keep separate, there's a lot of nuance there.

> I also noticed a lack of focus styles. Do we have CSS/styles that I can reuse for this?

There are a couple of mixins, but I'm curious why the focus style doesn't work on its own? Ideally it should be built into the button component itself.

**Comment by Mamaduka** [comment]
The `primary` button has a focus style when it's disabled. But not `secondary` or `tertiary`.

## Screenshot
CleanShot 2023-09-22 at 14 27 42

**Comment by jasmussen** [comment]
That seems a bug in the core component, one to fix there, no?

**Comment by Mamaduka** [comment]
Yes, it's unrelated to the changes in this PR, so we can fix it separately and add a changelog entry to the components package.

**Comment by alexstine** [comment]
Its a design challenge to solve Gutenberg-wide. There are 2 basic rules.
1. Avoid removing/re-rendering DOM elements.
2. Avoid changing the labels/accessible names of buttons to communicate state.

The second is less obvious but still a WCAG failure. The first can lead to focus loss as this PR shows.

What someone should figure out how to do is code some type of focus loss handler that always returns focus back to a constant point. Tricky part would be figuring out if its a true focus loss vs. the user moving away. At least until design and functionality learn to work together a little better in the UI. I tried it once but could never figure out how to tell if focus was totally lost or not. I have an active PR where I try to implement this for writing flow and I almost get a headache from how messy the code is.

Real solution, do whatever you've got to do visually but never change the DOM to the point where React can't reconcile.

**Comment by jameskoster** [comment]
Would it be feasible to use the status selection UI from the site editor here?

**Comment by Mamaduka** [comment]
@jameskoster, can you share the screenshot?

**Comment by jameskoster** [comment]
This one:

Screenshot 2023-09-25 at 13 51 21

To be clear, I'm not suggesting we do that in _this_ PR. Just noting that the site editor UI effectively solves this issue by omitting the "Switch to draft" button altogether. And since that UI already exists, perhaps it can be reused?

**Comment by Mamaduka** [comment]
Thanks, @jameskoster!

The post editor already has a similar UI, so we should consolidate Status control.

Does anyone remember the historical reasons why we had a separate button for the "Switch to Draft" action?

## Screenshot
CleanShot 2023-09-25 at 16 59 10

**Comment by afercia** [comment]
> Would it be feasible to use the status selection UI from the site editor here?

The site editor 'status selection' UI actually mixes together Status and Visibility, which historically are two separate logical concepts. I'm not sure why they were grouped together to start with. Seems to me that's less than ideal and i'd recommend to keep those concepts logically and visually separated.

Also, the status selection UI introduced a pretty visible inconsistency with the post editor, see https://github.com/WordPress/gutenberg/issues/52643. I'm not sure how this kind of UI onconsistencies can be of any help for users. 

There is a proposal to better align some edit features in the Site and Post editors, see https://github.com/WordPress/gutenberg/issues/52632 but still I'm not sure that mixing together status and visibility is the right way to go.

**Comment by jameskoster** [comment]
Whether status and visibility are separate or not, the point was to use the same UI in both locations. But clearly that is distracting from this PR, apologies @Mamaduka!

> Does anyone remember the historical reasons why we had a separate button for the "Switch to Draft" action?

I do not, but I remember it used to live in the main top bar, and was moved to reduce prominence: https://github.com/WordPress/gutenberg/pull/50217. 

Perhaps we pursue this PR as-is, then follow-up to try and better align post & site editor document inspectors. I'd be happy to work on designs for that.

**Comment by Mamaduka** [comment]
> Perhaps we pursue this PR as-is, then follow-up to try and better align post & site editor document inspectors. I'd be happy to work on designs for that.

Sounds good to me! Let's solve the immediate a11y issue; ensure the new design pattern we implement has a11y reviews before merging.

If there are no other blockers for this PR, I would appreciate approval 🙇

**Comment by jameskoster** [comment]
The blue outline on hover looks strange. 

Screenshot 2023-09-25 at 15 51 31

Should we fix that here or in a follow-up that targets the button component? It doesn't appear if the button is disabled rather than aria-disabled.

**Comment by Mamaduka** [comment]
@jameskoster, the `secondary` button also lacks a `focus` outline when disabled. It would be nice to fix button-related issues in a single follow-up. See also https://github.com/WordPress/gutenberg/pull/54722#issuecomment-1731188832

**Review by alexstine — APPROVED** [review]
@Mamaduka Thanks, works as described!

**Review by jameskoster — APPROVED** [review]
In that case let's land this then follow up to fix the Button issues :)

**Comment by Mamaduka** [comment]
@jameskoster, @jasmussen, do you have a mockup of how the disabled secondary button's focus/hover styles should look like? I can create a follow-up PR this week.

**Comment by jameskoster** [comment]
Thanks @Mamaduka :) 

Looking at [Storybook](https://wordpress.github.io/gutenberg/?path=/docs/components-button--docs) everything seems to work as expected, IE there are no hover or focus styles on disabled buttons. I'm not sure why that was creeping in here 🤔

**Comment by Mamaduka** [comment]
Try settings `__experimentalIsFocusable` and `disabled` simultaneously for the button in Storybook. I'll push PR shortly to fix the hover style.

I think @jasmussen said that the disabled button still should have a border - https://github.com/WordPress/gutenberg/pull/54722#issuecomment-1731152219.

## Screenshot

CleanShot 2023-10-02 at 19 43 57

**Comment by jameskoster** [comment]
> I think @jasmussen said that the disabled button still should have a border

I don't have a strong feeling about that, but I will say that with a gray border / text the button begins to resemble an input field which isn't ideal:

Screenshot 2023-10-02 at 17 14 18

If we want to maintain the boxed shape then a background might work better. This more closely resembles the default browser style for disabled buttons too.

Screenshot 2023-10-02 at 17 16 12

Is `__experimentalIsFocusable` applied here so that the button remains focussed (and becomes disabled) after converting to draft?

If that's required then I expect the standard outline makes sense: 

Screenshot 2023-10-02 at 17 19 21

**Comment by Mamaduka** [comment]
> Is __experimentalIsFocusable applied here so that the button remains focussed (and becomes disabled) after converting to draft?

Not via this prop. The component uses `aria-disabled` directly, which is emulated by the Button component when both `__experimentalIsFocusable` and `disabled` are set to `true`.

I've created https://github.com/WordPress/gutenberg/pull/54978, which matches the rules for other buttons.

**Comment by jasmussen** [comment]
If the button is disabled, the border does not need to meet any contrast ratios, it just needs to be focusable. So this is what I was thinking:
Sketch

The gray background lends prominence in this context which, wouldn't you agree, is unintentional for a disabled button?

Sketch

**Comment by Mamaduka** [comment]
> If the button is disabled, the border does not need to meet any contrast ratios, it just needs to be focusable. So this is what I was thinking:

@jameskoster, basically use the same grey border when a disabled button is focused, right?

**Comment by jameskoster** [comment]
The light border could work. Worth double-checking about the text colour though. Do disabled button labels still need to meet contrast guidelines?

**Comment by Mamaduka** [comment]
> Do disabled button labels still need to meet contrast guidelines?

If I had to guess, I would say yes 😅  But I would defer to @afercia, @alexstine, and @joedolson on this one.

**Comment by jasmussen** [comment]
[No, disabled buttons do not need to meet contrast ratios](https://www.w3.org/TR/WCAG21/#contrast-minimum). I'm happy to be educated if I'm wrong:

>  Incidental: Text or images of text that are part of an **inactive user interface component**, that are pure decoration, that are not visible to anyone, or that are part of a picture that contains significant other visual content, have no contrast requirement.

Emphasis mine.

**Comment by alexstine** [comment]
Looks like not.

> Inactive User Interface Components

https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html

**Comment by Mamaduka** [comment]
Thanks, everyone!

@jameskoster, which light border/box shadow can I use here? Any existing example would be helpful.

**Comment by jasmussen** [comment]
You can use the one I used, which is $gray-500 text and $gray-200 border. It would be good to do that too, this currently feels like it's floating a bit:

Screenshot 2023-10-05 at 10 06 10

**Comment by richtabor** [comment]
> You can use the one I used, which is $gray-500 text and $gray-200 border. It would be good to do that too, this currently feels like it's floating a bit:

Was coming to report the same; looks a bit broken.

**Comment by Mamaduka** [comment]
I will push fix for this today. Just want to clarify one thing. We want secondary disable button to always have a border not only when it’s focused. Is this correct assumption?

**Comment by jasmussen** [comment]
Secondary buttons should always have borders, whether focused or not, disabled or not.

**Comment by Mamaduka** [comment]
I'm a bit more confused now 😅 Do you mind if I leave this follow-up to you folks?

**Pull request merged**
